core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessageDuplicator.java

```diff
@@ -126,246 +128,128 @@ public void close() {
         private static final AtomicLongFieldUpdater<StreamMessageProcessor> requestedDemandUpdater =
                 AtomicLongFieldUpdater.newUpdater(StreamMessageProcessor.class, "requestedDemand");
 
-        private final StreamMessage<? extends T> upstream;
-
-        private final Set<DownstreamSubscription> downstreamSubscriptions =
+        private final StreamMessage<T> upstream;
+        private final Set<DownstreamSubscription<T>> downstreamSubscriptions =
                 Collections.newSetFromMap(new ConcurrentHashMap<>());
 
-        private final List<T> contentList = new ArrayList<>();
+        private final List<Object> signals = new ArrayList<>();
 
         @SuppressWarnings("unused")
         private volatile long requestedDemand;
         private volatile Subscription upstreamSubscription;
-        private volatile boolean upstreamCompleted;
-        private volatile int upstreamOffset;
         private volatile State state = State.DUPLICABLE;
 
-        private Throwable upstreamCause;
-
-        StreamMessageProcessor(StreamMessage<? extends T> upstream) {
+        StreamMessageProcessor(StreamMessage<T> upstream) {
             this.upstream = upstream;
             upstream.subscribe(this, true);
-            notifyDownstreamWhenUpstreamCompleted(upstream);
         }
 
-        private void notifyDownstreamWhenUpstreamCompleted(StreamMessage<? extends T> upstream) {
-            upstream.closeFuture().whenComplete((unused, cause) -> {
-                upstreamCompleted = true;
-                if (cause != null) {
-                    upstreamCause = cause;
-                }
-                notifyDownstreams();
-            });
+        StreamMessage<T> upstream() {
+            return upstream;
+        }
+
+        List<Object> signals() {
+            return signals;
         }
 
         @Override
         public void onSubscribe(Subscription s) {
             upstreamSubscription = s;
-            downstreamSubscriptions.forEach(downstream -> {
-                if (downstream.setSubscribed()) {
-                    if (downstream.executor() != null) {
-                        downstream.executor().execute(() -> downstream.subscriber().onSubscribe(downstream));
-                    } else {
-                        downstream.subscriber().onSubscribe(downstream);
-                    }
-                }
-            });
+            downstreamSubscriptions.forEach(DownstreamSubscription::invokeOnSubscribe);
+
             if (state == State.CLOSED) {
-                s.cancel();
+                upstream.abort();
```

Comment:
IIUC, there's only one way to change state `DUPLICABLE` to `CLOSED` that calling StreamMessageProcessor#close. And it might already call `upstream.abort()`. Do we need to call `upstream.abort()` here again?